### PR TITLE
render code path improvements

### LINKED ISSUE #2137: Move data sources on to GlyphRenderers
Right now every glyph has its own data source, which is very inefficient when there are also `selection_glyph`/`nonselectiong_glyph` (and later `inspection_glyph`) glyphs managed by the `GlyphRenderer`. Change so that all glyphs managed by a `GlyphRenderer` use a single data source on the `GlyphRenderer` instead.

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #2137

This PR cleans up and improves the render code path in several ways:

* greatly reduces unnecessary code and compute duplication among the glyphs of a given `GlyphRenderer`
* finer grained trace level logging for performance timing
* faster selection masking using a hash

A few numbers: On my machine for the `color_scatter.py` example with `N=40000` (forty thousand) these changes reduce the `set_data` time from 160ms to 60ms. Selection masking times were reduced by an order of magnitude or more, to single digit numbers of milliseconds. It was hard to measure, but I believe there was a ~30 percent reduction in JS stack allocation size. At this point for large N the render code path is dominated by canvas calls, which we can only fix by offering a different render backend (WebGL investigations underway)

### CODE REVIEW

**Comment by bryevdv** [comment]
@bokeh/core this is ready for review.

**Comment by pzwang** [comment]
YAY!!!

File: bokehjs/src/coffee/renderer/glyph/glyph_renderer.coffee [hunk 1]
```diff
@@ -48,73 +48,81 @@ class GlyphRendererView extends PlotWidget
...
-    else
+      tselect = Date.now()
       # reset the selection mask
-      selected_mask = (false for i in @all_indices)
-      for idx in selected_indices
-        selected_mask[idx] = true
+      selected_mask = {}
+      for i in selected
+        selected_mask[i] = true
+      #selected_mask = (i in selected for i in @all_indices)
```
Comment (fpliger) [inline bokehjs/src/coffee/renderer/glyph/glyph_renderer.coffee]: Forgot to delete this one?

**Comment by fpliger** [comment]
Nice imprevements. Really like the change to pass data more explicitly. LGTM

**Comment by bryevdv** [comment]
@bokeh/core any objections to merging this?

**Comment by fpliger** [comment]
:+1:

**Pull request merged**